File: src/app/dockWave.ts
/**
 * The dock icon: the station mark, drawn once and handed to the Dock at
 * boot. A dev binary has no bundle icon behind it - without this macOS shows
 * the generic executable tile - so the mark is rendered on a canvas from the
 * same silhouette the in-app wave uses and shipped over the bridge as PNG
 * bytes for AppKit to wear.
 *
 * This used to be a live meter that streamed spectrum-driven frames while
 * music played; the frames could never sit close enough to the beat to feel
 * synced, so the icon is still again.
 *
 * Everything here no-ops outside the Tauri webview.
 */

import { isTauri } from './tauri.ts';
import { sampleWave, WAVE_MID_Y, WAVE_SHADOW_DROP } from './BeatWave.tsx';

/** Rendered icon edge, in px. The Dock displays ~64-128; retina-friendly. */
const ICON_SIZE = 256;

/** The asset's stroke, as a share of its width, scaled to the canvas. */
const STROKE = 4.6 * (ICON_SIZE / 100);

/** The mark's two lines, straight off the asset. */
const WAVE_COLOR = '#ffffff';
const TRACER_COLOR = '#8a8a8a';

let sentStill = false;

/** The brand mark on black: the grey shadow line under the white wave. */
function drawStill(): Promise<Uint8Array | null> {
  const canvas = document.createElement('canvas');
  canvas.width = ICON_SIZE;
  canvas.height = ICON_SIZE;
  const ctx = canvas.getContext('2d');
  if (!ctx) return Promise.resolve(null);

  const scale = ICON_SIZE / 100;
  ctx.fillStyle = '#000';
  ctx.fillRect(0, 0, ICON_SIZE, ICON_SIZE);
  ctx.lineCap = 'round';
  ctx.lineJoin = 'round';
  ctx.lineWidth = STROKE;

  const base = sampleWave(1);
  const stroke = (drop: number, color: string) => {
    ctx.strokeStyle = color;
    ctx.beginPath();
    base.forEach(([x, dy], index) => {
      const px = x * scale;
      const py = (WAVE_MID_Y + drop + dy) * scale;
      if (index === 0) ctx.moveTo(px, py);
      else ctx.lineTo(px, py);
    });
    ctx.stroke();
  };

  stroke(WAVE_SHADOW_DROP, TRACER_COLOR);
  stroke(0, WAVE_COLOR);

  return new Promise((resolve) => {
    canvas.toBlob(async (blob) => {
      if (!blob) return resolve(null);
      resolve(new Uint8Array(await blob.arrayBuffer()));
    }, 'image/png');
  });
}

/** Renders and ships the brand frame. Called once at boot. */
export async function initDockWave(): Promise<void> {
  if (!isTauri() || sentStill) return;
  sentStill = true;
  const still = await drawStill();
  if (!still) {
    sentStill = false;
    return;
  }
  try {
    const { invoke } = await import('@tauri-apps/api/core');
    await invoke('dock_wave_still', { still: Array.from(still) });
  } catch {
    // A dock left generic is not worth failing boot over.
  }
}

File: src-tauri/src/dock_wave.rs
//! The dock icon.
//!
//! The webview draws the station mark once at boot and ships it here as raw
//! PNG bytes; this side hands it to the Dock. A dev binary has no bundle
//! icon behind it, so without this macOS shows the generic executable tile.
//! The icon is never handed back to macOS (`setApplicationIconImage(nil)`)
//! for the same reason.
//!
//! This used to receive a ~15fps stream of spectrum-driven frames while
//! music played; the meter never sat close enough to the beat to feel
//! synced, so the icon is still again.
//!
//! Everything here is a no-op off macOS: the command exists everywhere so
//! the frontend never needs to know, but only AppKit has a dock tile.

use tauri::AppHandle;

/// Replaces the dock icon. Main thread only - AppKit's rule - so the swap
/// is queued rather than performed in place.
#[cfg(target_os = "macos")]
fn set_dock_icon(app: &AppHandle, png: Vec<u8>) {
    let _ = app.run_on_main_thread({
        move || {
            use objc2::{AnyThread, MainThreadMarker};
            use objc2_app_kit::{NSApplication, NSImage};
            use objc2_foundation::NSData;
            let Some(mtm) = MainThreadMarker::new() else { return };
            let ns_app = NSApplication::sharedApplication(mtm);
            let data = NSData::with_bytes(&png);
            if let Some(image) = NSImage::initWithData(NSImage::alloc(), &data) {
                unsafe { ns_app.setApplicationIconImage(Some(&image)) };
            }
        }
    });
}

#[cfg(not(target_os = "macos"))]
fn set_dock_icon(_app: &AppHandle, _png: Vec<u8>) {}

/// Receives the brand frame and wears it. Sent at boot; a reload sends it
/// again, which just repaints the same mark.
#[tauri::command]
pub fn dock_wave_still(app: AppHandle, still: Vec<u8>) -> Result<(), String> {
    if still.is_empty() {
        return Err("dock icon needs a frame".into());
    }
    set_dock_icon(&app, still);
    Ok(())
}
